crates/engine/tree/src/tree/mod.rs

```diff
@@ -2887,20 +2887,23 @@ where
                         state_provider.state_root_with_updates(hashed_state.clone())?;
 
                     if regular_root == sealed_block.header().state_root() {
-                        let provider_ro = state_root_task_config.consistent_view.provider_ro()?;
-                        let in_memory_trie_cursor = InMemoryTrieCursorFactory::new(
-                            DatabaseTrieCursorFactory::new(provider_ro.tx_ref()),
-                            &state_root_task_config.nodes_sorted,
-                        );
-                        compare_trie_updates(
-                            in_memory_trie_cursor,
-                            task_trie_updates.clone(),
-                            regular_updates,
-                        )
-                        .map_err(ProviderError::from)?;
-                    } else {
-                        debug!(target: "engine::tree", "Regular state root does not match block state root");
+                        if self.config.always_compare_trie_updates() {
+                            let provider_ro =
+                                state_root_task_config.consistent_view.provider_ro()?;
+                            let in_memory_trie_cursor = InMemoryTrieCursorFactory::new(
+                                DatabaseTrieCursorFactory::new(provider_ro.tx_ref()),
+                                &state_root_task_config.nodes_sorted,
+                            );
+                            compare_trie_updates(
+                                in_memory_trie_cursor,
+                                task_trie_updates,
+                                regular_updates.clone(),
+                            )
+                            .map_err(ProviderError::from)?;
+                        }
+                        return Ok((regular_root, regular_updates, time_from_last_update))
```

Comment:
this is a bit weird now given that if `always_compare_trie_updates` is true and the state root matches, we would return hash builder output, but it's fine since they are supposed to be equivalent

Reply:
I would say it's important to use state root task outputs here, because that's what will be saved into the database.

Comparing updates only logs a warning, not stops the node, so if we want to test that sparse trie outputs result in a correct DB, and also compare the updates, we need to save state root task outputs here.

Reply:
@shekhirin ok, now it only returns the sequential results on mismatch and sequential results correct, otherwise returns the results from the tasks, PTAL

also pls take a look at the current behavior wrt `compare_trie_updates`, previously it was always called when there was a mismatch or we had the config set, now after this comment https://github.com/paradigmxyz/reth/pull/14387#pullrequestreview-2608109818 is only shown when the config is set.